### ISSUE TITLE
amp-social-share: Image-based buttons with missing text descriptions / accessible name

### ISSUE THREAD

**Issue by TetraLogicalHelpdesk** [ampproject/amphtml#29563]
## WCAG Level
Level A

## Priority
Critical

## Type
Images, Semantics

## Pages/Components affected
* [amp-social-share](https://amp.dev/documentation/components/amp-social-share/)

## Description
When images do not have a text description it means anyone who cannot see images properly will not know what the images are for. If the images act as actual controls (such as image-based buttons), this leads to the controls also lacking an accessible name. Without an accessible name, interactive controls will simply be announced as generic "link", "button" or similar by assistive technologies.

Currently, none of the "pre-configured providers" graphical share buttons have a text alternative/accessible name. Taking the [playground example](https://playground.amp.dev/?url=https%3A%2F%2Fpreview.amp.dev%2Fdocumentation%2Fexamples%2Fcomponents%2Famp-social-share&format=websites&_gl=1*1v8hyfe*_ga*YW1wLWdDQmZ0cDgwZl96MWxoRWR1NEFlaXc.), all the variations of the share buttons presented are simply exposed as generic "button" to assistive technologies. The definition of the `<amp-social-share>` element itself does not provide any default attribute to provide a text alternative/accessible name.

Among the "pre-configured providers" graphics, the "Line" graphic also contains the actual text "Line". As its accessible name (which is currently empty) does not currently contain this visible text, that particular item also fails WCAG 2.1 SC 2.5.3 Label in Name.

## User impact
If an image does not have an appropriate text description, someone who is blind and uses a screen reader or someone who has images turned off in their browser to save on bandwidth will be completely unaware of the image's purpose. If the image conveys important information like a graph, or is critical to the operation of a control like a link or button, then a missing text description can prevent them from completing vital tasks.

In the specific case of the "Line" control, speech-input users who normally use the visible text label as a means of navigation or selection will not be able to do so in this case - i.e. they won't be able to trigger the control by announcing "Press 'Line' button" or similar. Sighted users who use screen readers and who have some usable sight may be further confused when the text they hear (which currently would just be announced as an unlabelled "button") is different to the text that is visibly rendered on the screen ("Line").

## Required solution
The `<amp-social-share>` element must provide an text alternative/accessible name. Currently, an element defined as `amp-social-share type="email"></amp-social-share>` results in a rendered output of:

```
<amp-social-share type="email" class="i-amphtml-element i-amphtml-layout-fixed i-amphtml-layout-size-defined amp-social-share-email i-amphtml-layout" i-amphtml-layout="fixed" style="width: 60px; height: 44px;" role="button" tabindex="0"></amp-social-share>
```

The simplest way of providing an accessible name/alternative text for this element would be to add an `aria-label="..."` attribute, which will be used exposed as the accessible name for the button. The value of the attribute must appropriately reflect the purpose of the share button - at the very least, it should include information about the type (e.g. `aria-label="Email"`), but it can also be more explicit (e.g. `aria-label="Share by email"`), provided that this is not too verbose.

```
<amp-social-share type="email" class="..." i-amphtml-layout="fixed" style="..." role="button" tabindex="0" aria-label="Share by email"></amp-social-share>
```

For the pre-configured providers, this should be generated automatically without the author having to do anything specific in their AMP markup. Ideally, the AMP markup itself should also account for authors wanting to override the default accessible name for the pre-configured providers, as well as providing a standard attribute for authors to define the accessible name for "non-configured providers". While it is currently possible for authors to work around the limitation of the `<amp-social-share>` element by explicitly adding their own `aria-label`, we'd recommend adding adding a standard attribute to the element definition itself, such as `name` or `label`, which would then be turned into an `aria-label` when the element is rendered.

## Test procedure(s)
Use these steps to confirm that the solution has been correctly applied:
1. Turn on your screen reader
2. Navigate through all controls (using `TAB`/`SHIFT`+`TAB` to move to all focusable elements)
3. Make sure that controls are announced with an actual name and role (e.g. "Share by Email, button", "Share on Facebook, button", etc.) rather than just as generic nameless controls (e.g. "button")

## Related standards
* [WCAG 2.1 Success Criterion 1.1.1 Non-text Content (Level A)](https://www.w3.org/TR/WCAG21/#non-text-content)
* [WCAG 2.1 Success Criterion 2.5.3 Label in Name (Level A)](https://www.w3.org/TR/WCAG21/#label-in-name)
* [WCAG 2.1 Success Criterion 4.1.2 Name, Role, Value (Level A)](https://www.w3.org/TR/WCAG21/#name-value)

## More information
* [W3C Accessible Name and Description Computation 1.1](https://www.w3.org/TR/accname-1.1/)
* [W3C: Image concepts](https://www.w3.org/WAI/tutorials/images/decision-tree/)
* [W3C: An alt decision tree](https://www.w3.org/WAI/tutorials/images/decision-tree/)
* [W3C: HTML 5.2 Requirements for providing text to act as an alternative for images](https://www.w3.org/TR/html52/semantics-embedded-content.html#alt-text)

## Test data
Test date: July 2020
